File: pipeline/models/ships.py
"""Pydantic model for ships.json."""

from __future__ import annotations

from .common import LenientModel, StrictModel


class ConsoleSlots(StrictModel):
    """Console slot allocation for a hull."""

    arcana: int
    command: int
    engineering: int
    tactical: int
    universal: int


class WeaponSlots(StrictModel):
    """Weapon slot allocation for a hull."""

    forward: int
    rear: int


class Hull(StrictModel):
    """A ship hull."""

    id: str
    name: str
    # "class" is a Python keyword but valid JSON key — alias it
    hullClass: str | None = None
    cost: int
    crew: int
    hullStrength: int
    maneuverability: int
    acceleration: int
    speed: int
    sensors: int
    consoles: ConsoleSlots
    weapons: WeaponSlots

    class Config:
        # Allow the JSON field "class" to populate hullClass
        populate_by_name = True

    def model_post_init(self, __context: object) -> None:
        """Handle the 'class' field from JSON."""
        # Pydantic v2 doesn't auto-map unaliased 'class' to 'hullClass'
        # This is handled in the custom loader instead.


class Console(StrictModel):
    """A ship console component."""

    id: str
    name: str
    type: str
    cost: int
    effect: str


class ShipWeapon(LenientModel):
    """A ship weapon."""

    id: str
    name: str
    size: str
    material: str
    damage: str
    disruption: int
    accuracy: int
    crit: int | str
    range: int | str
    cost: int
    arc: str
    type: str


class Torpedo(LenientModel):
    """A torpedo type."""

    id: str
    name: str
    damage: str
    disruption: int
    accuracy: int
    crit: int | str
    arc: str
    range: int | str
    cost: int
    effect: str


class CriticalDamageEntry(StrictModel):
    """An entry on the ship critical damage table."""

    roll: str
    name: str
    effect: str


class Shield(LenientModel):
    """A ship shield."""

    id: str
    name: str
    type: str
    mark: int
    capacity: int
    regeneration: int
    special: str
    cost: int


class ShipsFile(LenientModel):
    """Top-level structure of ships.json."""

    holdingsBP: list[int]
    crewQualityCost: dict[str, int]
    hulls: list[Hull]
    consoles: list[Console]
    weapons: list[ShipWeapon]
    torpedoTubeCost: int
    torpedoes: list[Torpedo]
    shields: list[Shield]
    criticalDamage: list[CriticalDamageEntry]

File: src/lib/dtd/schemas/ships.ts
/**
 * Zod schema for ships data — replaces former Pydantic model.
 */

import { z } from "zod";

// ---------------------------------------------------------------------------
// ConsoleSlots (strict)
// ---------------------------------------------------------------------------

export const ConsoleSlots = z
	.object({
		arcana: z.number().int(),
		command: z.number().int(),
		engineering: z.number().int(),
		tactical: z.number().int(),
		universal: z.number().int(),
	})
	.strict();
export type ConsoleSlots = z.infer<typeof ConsoleSlots>;

// ---------------------------------------------------------------------------
// WeaponSlots (strict)
// ---------------------------------------------------------------------------

export const WeaponSlots = z
	.object({
		forward: z.number().int(),
		rear: z.number().int(),
	})
	.strict();
export type WeaponSlots = z.infer<typeof WeaponSlots>;

// ---------------------------------------------------------------------------
// Hull (strict)
// ---------------------------------------------------------------------------

/** The JSON field is "class" — Python aliases it to hullClass because class is reserved. */
export const Hull = z
	.object({
		id: z.string(),
		name: z.string(),
		class: z.string().nullish(),
		cost: z.number().int(),
		crew: z.number().int(),
		hullStrength: z.number().int(),
		maneuverability: z.number().int(),
		acceleration: z.number().int(),
		speed: z.number().int(),
		sensors: z.number().int(),
		consoles: ConsoleSlots,
		weapons: WeaponSlots,
	})
	.strict();
export type Hull = z.infer<typeof Hull>;

// ---------------------------------------------------------------------------
// Console (strict)
// ---------------------------------------------------------------------------

export const Console = z
	.object({
		id: z.string(),
		name: z.string(),
		type: z.string(),
		cost: z.number().int(),
		effect: z.string(),
	})
	.strict();
export type Console = z.infer<typeof Console>;

// ---------------------------------------------------------------------------
// ShipWeapon (lenient)
// ---------------------------------------------------------------------------

export const ShipWeapon = z
	.object({
		id: z.string(),
		name: z.string(),
		size: z.string(),
		material: z.string(),
		damage: z.string(),
		disruption: z.number().int(),
		accuracy: z.number().int(),
		crit: z.union([z.number().int(), z.string()]),
		range: z.union([z.number().int(), z.string()]),
		cost: z.number().int(),
		arc: z.string(),
		type: z.string(),
	})
	.passthrough();
export type ShipWeapon = z.infer<typeof ShipWeapon>;

// ---------------------------------------------------------------------------
// Torpedo (lenient)
// ---------------------------------------------------------------------------

export const Torpedo = z
	.object({
		id: z.string(),
		name: z.string(),
		damage: z.string(),
		disruption: z.number().int(),
		accuracy: z.number().int(),
		crit: z.union([z.number().int(), z.string()]),
		arc: z.string(),
		range: z.union([z.number().int(), z.string()]),
		cost: z.number().int(),
		effect: z.string(),
	})
	.passthrough();
export type Torpedo = z.infer<typeof Torpedo>;

// ---------------------------------------------------------------------------
// CriticalDamageEntry (strict)
// ---------------------------------------------------------------------------

export const CriticalDamageEntry = z
	.object({
		roll: z.string(),
		name: z.string(),
		effect: z.string(),
	})
	.strict();
export type CriticalDamageEntry = z.infer<typeof CriticalDamageEntry>;

// ---------------------------------------------------------------------------
// Shield (lenient)
// ---------------------------------------------------------------------------

export const Shield = z
	.object({
		id: z.string(),
		name: z.string(),
		type: z.string(),
		mark: z.number().int(),
		capacity: z.number().int(),
		regeneration: z.number().int(),
		special: z.string(),
		cost: z.number().int(),
	})
	.passthrough();
export type Shield = z.infer<typeof Shield>;

// ---------------------------------------------------------------------------
// ShipsFile (lenient)
// ---------------------------------------------------------------------------

export const ShipsFile = z
	.object({
		holdingsBP: z.array(z.number().int()),
		crewQualityCost: z.record(z.string(), z.number().int()),
		hulls: z.array(Hull),
		consoles: z.array(Console),
		weapons: z.array(ShipWeapon),
		torpedoTubeCost: z.number().int(),
		torpedoes: z.array(Torpedo),
		shields: z.array(Shield),
		criticalDamage: z.array(CriticalDamageEntry),
	})
	.passthrough();
export type ShipsFile = z.infer<typeof ShipsFile>;
